File: legacy-python/data_structures/reputation.py
from __future__ import annotations

from collections import defaultdict
from typing import Dict


class ReputationTracker:
    """Track member actions and update reputation based on events."""

    def __init__(self, dao, decay_rate: float = 0.01) -> None:
        self.dao = dao
        self.decay_rate = decay_rate
        self.last_activity: Dict[str, int] = defaultdict(lambda: dao.current_step)
        if getattr(dao, "event_bus", None):
            dao.event_bus.subscribe("project_worked", self._handle_event)
            dao.event_bus.subscribe("service_offered", self._handle_event)
            dao.event_bus.subscribe("proposal_invested", self._handle_event)

    def _get_member(self, uid: str):
        for m in self.dao.members:
            if m.unique_id == uid:
                return m
        return None

    def _handle_event(self, event: str, **data) -> None:
        if event == "project_worked":
            member = self._get_member(data.get("member"))
            if member is not None:
                work = data.get("work", 0)
                member.reputation += work / 10
                self.last_activity[member.unique_id] = self.dao.current_step
        elif event == "service_offered":
            member = self._get_member(data.get("provider"))
            if member is not None:
                member.reputation += 1
                self.last_activity[member.unique_id] = self.dao.current_step
        elif event == "proposal_invested":
            member = self._get_member(data.get("investor"))
            if member is not None:
                amount = data.get("amount", 0)
                member.reputation += amount / 100
                self.last_activity[member.unique_id] = self.dao.current_step

    def decay_reputation(self) -> None:
        """Apply inactivity decay based on ``DAO.current_step``."""
        for member in self.dao.members:
            last = self.last_activity.get(member.unique_id, 0)
            if self.dao.current_step > last:
                member.reputation *= 1 - self.decay_rate

File: lib/data-structures/reputation.ts
// Reputation Tracker - tracks member actions and updates reputation
// Port from data_structures/reputation.py

import type { DAO } from './dao';
import type { DAOMember } from '../agents/base';

export class ReputationTracker {
  dao: DAO;
  decayRate: number;
  lastActivity: Map<string, number> = new Map();

  constructor(dao: DAO, decayRate: number = 0.01) {
    this.dao = dao;
    this.decayRate = decayRate;

    // Subscribe to relevant events
    if (dao.eventBus) {
      dao.eventBus.subscribe('project_worked', this.handleEvent.bind(this));
      dao.eventBus.subscribe('service_offered', this.handleEvent.bind(this));
      dao.eventBus.subscribe('proposal_invested', this.handleEvent.bind(this));
    }
  }

  /**
   * Get member by unique ID
   */
  private getMember(uid: string): DAOMember | null {
    return this.dao.members.find((m) => m.uniqueId === uid) || null;
  }

  /**
   * Handle reputation-affecting events
   */
  private handleEvent(data: Record<string, any>): void {
    const currentStep = (this.dao as any).currentStep || 0;
    const event = data.event;

    switch (event) {
      case 'project_worked': {
        const member = this.getMember(data.member);
        if (member) {
          const work = data.work || 0;
          member.reputation += work / 10;
          this.lastActivity.set(member.uniqueId, currentStep);
        }
        break;
      }

      case 'service_offered': {
        const member = this.getMember(data.provider);
        if (member) {
          member.reputation += 1;
          this.lastActivity.set(member.uniqueId, currentStep);
        }
        break;
      }

      case 'proposal_invested': {
        const member = this.getMember(data.investor);
        if (member) {
          const amount = data.amount || 0;
          member.reputation += amount / 100;
          this.lastActivity.set(member.uniqueId, currentStep);
        }
        break;
      }
    }
  }

  /**
   * Apply inactivity decay to all members
   */
  decayReputation(): void {
    const currentStep = (this.dao as any).currentStep || 0;

    for (const member of this.dao.members) {
      const lastActive = this.lastActivity.get(member.uniqueId) || 0;

      if (currentStep > lastActive) {
        member.reputation *= 1 - this.decayRate;
      }
    }
  }

  /**
   * Get reputation statistics
   */
  getStatistics(): {
    average: number;
    median: number;
    min: number;
    max: number;
    totalAboveZero: number;
    totalBelowZero: number;
  } {
    if (this.dao.members.length === 0) {
      return {
        average: 0,
        median: 0,
        min: 0,
        max: 0,
        totalAboveZero: 0,
        totalBelowZero: 0,
      };
    }

    const reputations = this.dao.members.map((m) => m.reputation).sort((a, b) => a - b);
    const sum = reputations.reduce((s, r) => s + r, 0);
    const average = sum / reputations.length;
    const median = reputations[Math.floor(reputations.length / 2)];
    const min = reputations[0];
    const max = reputations[reputations.length - 1];
    const totalAboveZero = reputations.filter((r) => r > 0).length;
    const totalBelowZero = reputations.filter((r) => r < 0).length;

    return {
      average,
      median,
      min,
      max,
      totalAboveZero,
      totalBelowZero,
    };
  }

  /**
   * Get members ranked by reputation
   */
  getRankings(limit: number = 10): Array<{ id: string; reputation: number }> {
    return this.dao.members
      .map((m) => ({ id: m.uniqueId, reputation: m.reputation }))
      .sort((a, b) => b.reputation - a.reputation)
      .slice(0, limit);
  }
}
